src/Storages/ObjectStorage/DataLakes/IcebergMetadata.cpp

```diff
@@ -240,91 +235,203 @@ DataTypePtr getFieldType(const Poco::JSON::Object::Ptr & field, const String & t
+                        id,
+                        current_old_id,
+                        current_new_id);
+                }
+                else
                 {
-                    schema = current_schema;
-                    break;
+                    outputs.push_back(old_node);
                 }
             }
+            else
+            {
+                if (old_json->isObject("type"))
+                {
+                    throw Exception(
+                        ErrorCodes::UNSUPPORTED_METHOD,
+                        "Can't cast primitive type to the complex type, field id is {}, old schema id is {}, new schema id is {}",
+                        id,
+                        current_old_id,
+                        current_new_id);
+                }
+                String old_type = old_json->getValue<String>("type");
+                String new_type = field->getValue<String>("type");
 
-            if (!schema)
-                throw Exception(ErrorCodes::BAD_ARGUMENTS, R"(There is no schema with "schema-id" that matches "current-schema-id" in metadata)");
+                const ActionsDAG::Node * node = old_node;
+                if (old_type == new_type)
+                {
+                    if (old_json->getValue<String>("name") != name)
+                    {
+                        node = &dag->addAlias(*old_node, name);
+                    }
+                }
+                else if (allowPrimitiveTypeConversion(old_type, new_type))
+                {
+                    node = &dag->addCast(*old_node, getFieldType(field, "type", required), name);
+                }
+                outputs.push_back(node);
+            }
         }
         else
         {
-            if (schemas->size() != 1)
-                throw Exception(ErrorCodes::UNSUPPORTED_METHOD,
-                    "Cannot read Iceberg table: the table schema has been changed at least 1 time, reading tables with evolved schema is "
-                    "supported. If you want to ignore schema evolution and read all files using latest schema saved on table creation, enable setting "
-                    "iceberg_engine_ignore_schema_evolution (Note: enabling this setting can lead to incorrect result)");
-
-            /// Now we sure that there is only one schema.
-            schema = schemas->getObject(0);
-            if (schema->getValue<int>("schema-id") != current_schema_id)
-                throw Exception(ErrorCodes::BAD_ARGUMENTS, R"(Field "schema-id" of the schema doesn't match "current-schema-id" in metadata)");
+            if (field->isObject("type"))
+            {
+                throw Exception(
+                    ErrorCodes::UNSUPPORTED_METHOD,
+                    "Adding a default column with id {} and complex type is not supported yet, old schema id is {}, new schema id is {}",
+                    id,
+                    current_old_id,
+                    current_new_id);
+            }
```

Comment:
What is the problem with adding new columns with complex types?

Reply:
I encountered several problems (difference between nullability of struct and struct fields which for Parquet is sensible and CH is not, sth else, difficult logic for structs which are changed as an array element). I decided to postpone dealing with this problems to deliver some other features

Reply:
I discussed this with Sasha and Melvyn they are ok with it

Reply:
Got it